### PR TITLE
Long/show user status

### PR DESCRIPTION
**Pull request by khuolo01**
Ticket: https://tan-tho-le.atlassian.net/jira/software/projects/VAJ/boards/1?selectedIssue=VAJ-27

This PR implements a feature showing the user's status.

### CODE REVIEW

**Comment by leth02** [comment]
Can you provide a screenshot video of what the feature looks like?

From a quick look, I doubt that the feature would work, as you are setting a global clock (`LAST_ACTIVE_INTERVAL`) to detect when a user is last active. However, this clock will affect ALL users, not just the single user you are targeting.

Keep in mind that other objects in Flask are[ proxies to other actual objects](https://flask.palletsprojects.com/en/2.0.x/reqcontext/#notes-on-proxies) in the background, and this is a special mechanism of Flask being done with the use of [Werkzeug's Context Locals](https://werkzeug.palletsprojects.com/en/2.0.x/local/). Take `Session `for example. We just need to use `Session`, and Flask (Context Locals under the hood) will route the operation/data we pass to the corresponding object. We do not explicitly do that.

Here you are using a regular Python global variable. Does it really work?

**Comment by khuolo01** [comment]
> Can you provide a screenshot video of what the feature looks like?
> 
> From a quick look, I doubt that the feature would work, as you are setting a global clock (`LAST_ACTIVE_INTERVAL`) to detect when a user is last active. However, this clock will affect ALL users, not just the single user you are targeting.
> 
> Keep in mind that other objects in Flask are[ proxies to other actual objects](https://flask.palletsprojects.com/en/2.0.x/reqcontext/#notes-on-proxies) in the background, and this is a special mechanism of Flask being done with the use of [Werkzeug's Context Locals](https://werkzeug.palletsprojects.com/en/2.0.x/local/). Take `Session `for example. We just need to use `Session`, and Flask (Context Locals under the hood) will route the operation/data we pass to the corresponding object. We do not explicitly do that.
> 
> Here you are using a regular Python global variable. Does it really work?

The python global variable _LAST_ACTIVE_INTERVAL_ is a constant variable that stores the maximum time a user can close their browser yet still counts as "active" (currently 10 minutes, or 10 * 60 * 1000 (milliseconds)). The parts that use this global variable are:  

**_status = "active" if (time.time() * 1000 - last_active_time < LAST_ACTIVE_INTERVAL) else "away"_**
**_new_status = "active" if time.time() * 1000 - receiver.last_active_time < LAST_ACTIVE_INTERVAL else "away"_**

Both lines indicate "the status is active if the difference between the current time and the last_active_time is less than 10 minutes". The global variable LAST_ACTIVE_INTERVAL represents the number 10 minutes, and it is not used to set any clock. The reason I create the global variable is the interval time may be subject to change in the future. For example, we may think 10 minutes is too long, and we want to update the user's status to **_away_** after they close the browser for 5 minutes. If that is the case, we can just change the value of the variable from **600000** to **300000**.

The actual variables for setting the clock are _**conversation_container_refresh_interval**_ and **_update_last_active_time_interval_**. Both variables are in the javascript files. The former is used to set the clock to refresh the conversation container, and the latter is used to set the clock to refresh the last_active_time of an active user.

The video of how it functions I will send through slack.

File: message_app/send_message.py [hunk 1]
```diff
@@ -1,21 +1,50 @@
...
+        participants = conv.participants
+        if participants[0].username == current_user:
+            receiver_name = participants[1].username
+            last_active_time = participants[1].last_active_time
+        else:
+            receiver_name = participants[0].username
+            last_active_time = participants[0].last_active_time
+        conv_id = conv.id
+
+        # time.time() * 1000 because time function in Python returns time in seconds, while JS returns time in milliseconds
```
Comment (leth02) [inline message_app/send_message.py]: This line is not really necessary.

File: message_app/send_message.py [hunk 2]
```diff
@@ -1,21 +1,50 @@
...
+from flask import Blueprint, render_template, session
 from message_app.model import Messages, Conversations, User
-from message_app.db.db import DB as db
+from message_app.db.db import DB
 
 send_messages = Blueprint("send_messages", __name__)
 socketio = SocketIO(cors_allowed_origins='*')
 
+# An interval that decides user's status (ACTIVE or AWAY)
+# user's last_active_time < LAST_ACTIVE_INTERVAL means the user is active and vice versa
```
Comment (leth02) [inline message_app/send_message.py]: I understand its use here but the description is not so correct. You could change it to something like this.
```Python
# Define the amount of time a user can be active/inactive on the frontend before we actually update their status on the server
```

File: message_app/send_message.py [hunk 3]
```diff
@@ -1,21 +1,50 @@
...
+LAST_ACTIVE_INTERVAL = 10 * 60 * 1000 # 600000 milliseconds or 10 minutes
+
 @send_messages.route("/messages", methods=["GET"])
 def messages():
     # Get all conversations from an user
```
Comment (leth02) [inline message_app/send_message.py]: Get conversations of the current user

File: message_app/send_message.py [hunk 4]
```diff
@@ -36,6 +65,27 @@ def leaving(data):
...
+    for c in conversations:
+        receiver = User.select(c["receiver_name"])
+        new_status = "active" if time.time() * 1000 - receiver.last_active_time < LAST_ACTIVE_INTERVAL else "away"
+        c["conversation_status"] = new_status
+    emit("update_conversations_container", conversations)
+
+
+# ================== Sockets for Messages Container ==========================
+
 # A socket that checks if an user is typing
```
Comment (leth02) [inline message_app/send_message.py]: a user

File: message_app/static/js/conversations/main.js [hunk 5]
```diff
@@ -36,29 +38,78 @@ class ConversationHTMLElement {
...
-    },
-    {
-        id: 78645645,
-        title: "Luc Vuong, Khuong Long, and 1+ other people" // If more than two other users join this conversation, join their usernames together
-    }
-];
-
-for (const c of conversationArrays) {
+// Load conversation container when user accesses the messages page
+for (const c of available_conversations) {
```
Comment (leth02) [inline message_app/static/js/conversations/main.js]: please use CamelCase here: `availableConversations`

File: message_app/static/js/conversations/main.js [hunk 6]
```diff
@@ -36,29 +38,78 @@ class ConversationHTMLElement {
...
+        socket.emit("join", {"username": username, "conversation_id": conversation_id});
+        let message_panel = document.getElementById("message-panel");
+        message_panel.innerHTML = "You are in conversation with " + conversation_title;
+        conversation = new_conversation;
+    }
+}
+
+// Leave a conversation
+function leaveConversation(conversation_id){
+    socket.emit("leave", {"username": username, "conversation_id": conversation_id});
```
Comment (leth02) [inline message_app/static/js/conversations/main.js]: Update variable from `snake_case `to `camelCase`

File: message_app/static/js/socketio.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
+// Making the Websocket the only connection.
+const socket = io('http://127.0.0.1:5000/messages', { transports: ["websocket"] });
```
Comment (leth02) [inline message_app/static/js/socketio.js:2]: This should not be hard coded, but I will fix it later. Just a note.

File: message_app/static/js/socketio.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+// An event that receives messages from the server
+socket.on("message_handler_client", function(data){
+    if (!("join" in data)){
+        const message = new MessageModel(data.id, data.conversation_id, data.username, data.message, false, data.created_at);
+        FEEDBACK_HTML.innerHTML = "";
+        message.show();
+    }
+});
+
+// An event that shows if an user is typing
```
Comment (leth02) [inline message_app/static/js/socketio.js]: a user

File: message_app/templates/messages.html [hunk 9]
```diff
@@ -42,10 +37,11 @@
         <!-- Socket IO JS -->
         <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/4.0.1/socket.io.js" integrity="sha512-q/dWJ3kcmjBLU4Qc47E4A9kTB4m3wuTY7vkFJDTZKjTs8jhyGQnaUrxa0Ytd0ssMZhbNua9hE+E7Qv1j+DyZwA==" crossorigin="anonymous"></script>
```
Comment (leth02) [inline message_app/templates/messages.html]: I asked you to store this library on the server right?

**Review by leth02 — COMMENTED** [review]
Please resolve my comments here.

File: message_app/templates/messages.html (hunk 9 above)
Comment (khuolo01) [reply message_app/templates/messages.html]: Yes. I forgot about it. Just stored to the server in the last commit.

**Review by leth02 — APPROVED** [review]

**Pull request merged**